File: R/genome.r
#' read genome files (i.e., UCSC "chromSize" files)
#' 
#' Genome files contain chromosome name and size information. These sizes are 
#' used by downstream functions to identify intervals that have ends outside the
#' defined size.
#' 
#' @param filename file containing chrom/contig names and sizes, one-pair-per-line, 
#'   tab-delimited
#'   
#' @return \code{dplyr::tbl_df} with colnames \code{chrom} and \code{size},
#'   sorted by \code{chrom}
#'   
#' @examples
#' genome <- read_genome('hg19.chrom.sizes.gz')  
#' 
#' @export
read_genome <- function(filename) {
  colnames <- c('chrom', 'size')
  genome <- readr::read_tsv(filename, col_names = colnames)
  genome <- dplyr::tbl_df(genome) %>% dplyr::arrange(chrom)
  genome
}

File: src/genome.cpp
// genome.cpp
// 
// get genome information

#include <Rcpp.h>
using namespace Rcpp ;

std::map<std::string, int> create_genome(Rcpp::DataFrame genome) {
  
  std::vector <std::string> genome_chrom_v = 
    Rcpp::as< std::vector <std::string> >(genome["chrom"]); 
  std::vector <int> genome_size_v = 
    Rcpp::as< std::vector <int> >(genome["size"]); 
 
  std::map<std::string, int> out ;
  
  for (unsigned i = 0; i < genome_chrom_v.size(); ++i) {
    out[ genome_chrom_v[i] ] = genome_size_v[i] ; 
  }
 
  return out ; 
}
